compiler/src/compile.rs

```diff
@@ -127,16 +114,24 @@ pub fn compile_top(
     }
 }
 
-macro_rules! compile_impl {
-    ($ast:expr, $source_path:expr, $opts:expr, $st:ident, $compile:ident) => {{
-        let symbol_table = match $st($ast) {
-            Ok(x) => x,
-            Err(e) => return Err(e.into_compile_error($source_path)),
-        };
-        with_compiler($source_path, $opts, |compiler| {
-            compiler.$compile($ast, symbol_table)
-        })
-    }};
+/// A helper function for the shared code of the different compile functions
+fn compile_impl<Ast: ?Sized>(
+    ast: &Ast,
+    source_path: String,
+    opts: CompileOpts,
+    make_symbol_table: fn(&Ast) -> Result<SymbolTable, symboltable::SymbolTableError>,
+    compile: fn(&mut Compiler, &Ast, SymbolTable) -> CompileResult<()>,
```

Comment:
`fn()` pointers are dynamically dispatched, `impl Fn*` is monomorphized. It probably would be inlined anyway but it's generic anyway and that's kinda the idiomatic thing.

Reply:
The caller is a generic function already. it will inline the calling part regardless it is a `fn()` or `Fn()`. The called part will not be affect by them anyway. We even don't have `#[inline]` on top of `make_symbol_table` and `make_symbol_table_expr`.